### PR TITLE
Refactor component handling and generate shape information earlier.

### PR DESCRIPTION
**Pull request by schweitzpgi**
This merges all component handling back to a common path. Arrays, WHERE
constructs, and FORALL constructs all share the share common code to
derive the component path and any implicit iteration space shape
information. This refactoring causes generated code to appear in a
different order, so many tests were updated to reflect the changes. The
forall tests were split into individual tests.

  - Adds a component path collector to manage gathering and using a path
    of components.
  - Eliminates passing context information by this pointer, which
    fixes some bugs.
  - Fixes an off-by-one correctness bug.
  - Fixes some bugs with malformed iteration spaces, which resulted
    in assertion failures.
  - Fixes d7b abd d7c.

The todos left over:

  - Does not correctly deal with stemming the array properties relative
    to a path. This is a long standing issue with some hacks like
    componentToExtendedValue() that need to be replaced. This is related
    to the inability to handle component lower bounds, properly thread
    type parameters, etc.
  - Need to investigate a problem with degenerate slices, a(i:i).
  - There is still old prototype code to be removed.

### CODE REVIEW

File: flang/lib/Lower/ConvertExpr.cpp [hunk 1]
```diff
@@ -5467,285 +5356,331 @@ class ArrayExprLowering {
...
+                              // subs.
+                              auto subscriptVal = fir::getBase(asScalar(e));
+#if 0
+                              // arrayExv is the base array. It needs to reflect
+                              // the current array component instead.
+                              auto lb = fir::factory::readLowerBound(
+                                  builder, loc, arrayExv, ssIndex, one);
+#else
+                              auto lb = one;
+#endif
```
Comment (psteinfeld) [inline flang/lib/Lower/ConvertExpr.cpp:5424]: Is this left over from debugging something?

File: flang/lib/Lower/ConvertExpr.cpp [hunk 2]
```diff
@@ -5467,285 +5356,331 @@ class ArrayExprLowering {
...
+        return genImplicitArrayAccess(x, components);
+    }
+    bool atEnd = pathIsEmpty(components);
+    components.reversePath.push_back(&x);
+    auto result = genarr(x.base(), components);
+    if (components.applied)
       return result;
-    if (top)
+    if (atEnd) {
+      // components.reversePath.clear();
```
Comment (psteinfeld) [inline flang/lib/Lower/ConvertExpr.cpp:5615]: Should this line be deleted?

File: flang/lib/Lower/ConvertExpr.cpp [hunk 3]
```diff
@@ -5467,285 +5356,331 @@ class ArrayExprLowering {
...
-    if (top) {
-      reversePath.clear();
-      return [=](IterSpace) { return asScalar(x); };
-    }
     auto loc = getLoc();
-    return [=](IterSpace) {
-      fir::emitFatalError(loc, "QQ reached arrayref with path");
-      return ExtValue{};
+    if (atEnd) {
+      // components.reversePath.clear();
```
Comment (psteinfeld) [inline flang/lib/Lower/ConvertExpr.cpp:5653]: Should this line be deleted?

**Review by psteinfeld — CHANGES_REQUESTED** [review]
I had a few comments to be looked at.

Things built without problem, but when I ran `check-flang` I got errors.  Also, it would be good to have someone who understands the code better than I do look at it.
```
FAIL: Flang :: Lower/forall/forall-allocatable-2.f90 (315 of 1271)
******************** TEST 'Flang :: Lower/forall/forall-allocatable-2.f90' FAILED ********************
Script:
--
: 'RUN: at line 3';   bbc -emit-fir /local/home/psteinfeld/cg/eric/flang/test/Lower/forall/forall-allocatable-2.f90 -o - | /local/home/psteinfeld/cg/b1/build/bin/FileCheck /local/home/psteinfeld/cg/eric/flang/test/Lower/forall/forall-allocatable-2.f90
--
Exit Code: 1

Command Output (stderr):
--
/local/home/psteinfeld/cg/eric/flang/test/Lower/forall/forall-allocatable-2.f90:22:10: error: CHECK: expected string not found in input
! CHECK: %[[VAL_4:.*]] = fir.address_of(@_QQcl.ac2f54c8045ca4c820ddb680f252eda4) : !fir.ref<!fir.char<1,88>>
         ^
<stdin>:4:228: note: scanning from here
 %2 = fir.embox %1 : (!fir.ref<!fir.type<_QFforall_with_allocatable2Tt{i:i32,arr:!fir.box<!fir.heap<!fir.array<?xf32>>>}>>) -> !fir.box<!fir.type<_QFforall_with_allocatable2Tt{i:i32,arr:!fir.box<!fir.heap<!fir.array<?xf32>>>}>>
                                                                                                                                                                                                                                   ^
<stdin>:5:2: note: possible intended match here
 %3 = fir.address_of(@_QQcl.d58d6c1852f4da610d0afe6159ea251f) : !fir.ref<!fir.char<1,80>>
 ^

Input file: <stdin>
Check file: /local/home/psteinfeld/cg/eric/flang/test/Lower/forall/forall-allocatable-2.f90

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: func @_QPforall_with_allocatable2(%arg0: !fir.box<!fir.array<?xf32>>) { 
            2:  %0 = fir.alloca i32 {adapt.valuebyref, bindc_name = "i"} 
            3:  %1 = fir.alloca !fir.type<_QFforall_with_allocatable2Tt{i:i32,arr:!fir.box<!fir.heap<!fir.array<?xf32>>>}> {bindc_name = "thing", uniq_name = "_QFforall_with_allocatable2Ething"} 
            4:  %2 = fir.embox %1 : (!fir.ref<!fir.type<_QFforall_with_allocatable2Tt{i:i32,arr:!fir.box<!fir.heap<!fir.array<?xf32>>>}>>) -> !fir.box<!fir.type<_QFforall_with_allocatable2Tt{i:i32,arr:!fir.box<!fir.heap<!fir.array<?xf32>>>}>> 
check:22'0                                                                                                                                                                                                                                        X error: no match found
            5:  %3 = fir.address_of(@_QQcl.d58d6c1852f4da610d0afe6159ea251f) : !fir.ref<!fir.char<1,80>> 
check:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:22'1      ?                                                                                         possible intended match
            6:  %c5_i32 = arith.constant 5 : i32 
check:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            7:  %4 = fir.convert %2 : (!fir.box<!fir.type<_QFforall_with_allocatable2Tt{i:i32,arr:!fir.box<!fir.heap<!fir.array<?xf32>>>}>>) -> !fir.box<none> 
check:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            8:  %5 = fir.convert %3 : (!fir.ref<!fir.char<1,80>>) -> !fir.ref<i8> 
check:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            9:  %6 = fir.call @_FortranAInitialize(%4, %5, %c5_i32) : (!fir.box<none>, !fir.ref<i8>, i32) -> none 
check:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           10:  %c5_i32_0 = arith.constant 5 : i32 
check:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            .
            .
            .
>>>>>>

--

********************
```

**Comment by schweitzpgi** [comment]
Thanks for catching the directory hash problem, Pete.

The "#if 0" block should be live, but it doesn't always work because of the bug(s) mentioned in the description related to component paths. That is still TODO. I believe that the team has opted to not support components with arrays with lower bounds that are not 1, for the time being, so the "#else" block just assumes that is the case.

**Review by psteinfeld — APPROVED** [review]
All builds, tests, and looks good.

**Pull request merged**